### ISSUE TITLE
Cannot work when no rules are found

### ISSUE THREAD

**Issue by DexterZeng** [thunlp/MuGNN#2]
Hi, 

Thank you for providing the code! It has been really helpful!

However, when I tried to implement the code on a new dataset, where the number of relational triples is relatively small (3w triples for 1.5w entities), it failed to work. In specific, AMIE+ cannot find any rules, which are utilized in several places in the following codes (where the errors take place). Can you kindly let know how to modify the codes to make it work without rules? Or is it an indispensable component ?

Many thanks,
Weixin.

**Comment by DexterZeng** [comment]
Hi, it seems that I misunderstand the format of triples in AMIE folder (split by '\t' instead of ‘ ’). Now the codes can work well on the new dataset, for 5 epochs, and after evaluation, it reports an error like this:
![prob-1](https://user-images.githubusercontent.com/14938032/67581022-dc431c80-f779-11e9-8a4f-2ee50dc2e85c.jpg)

The printed numbers denote len(sim_chunked). Do you happen to know where might go wrong? 
P.S., due to the sparsity of relational triples and the small number of relations, AMIE discovers very few rules and the in the beginning, the number of new rules is 0 (shown in the figure below). Is this related to the error?

![prob-2](https://user-images.githubusercontent.com/14938032/67581384-7efb9b00-f77a-11e9-9b55-fcd01f758083.jpg)

Thank you for your time!

**Comment by acharkq** [comment]
Sorry for the late response and thank you for your attention to our work. 

For the assertion error, it seems that you are using a relatively small sim matrix in the `multi_process_get_nearest_neighbor` function. Try to change the variable `n_p` in `multi_process_get_nearest_neighbor` from 4 to 1 and see if the same error was raised

**Comment by acharkq** [comment]
Can you please print the shape of the `sim` matrix in `multi_process_get_nearest_neighbor`?

**Comment by DexterZeng** [comment]
Hi, thank you very much for your response.

The shape of entity sim matrix is (1779, 1779), while the shape of entity sim matrix is (3,3) (very few aligned relations). I have changed n_p to 1 and it reports error as below:

![error](https://user-images.githubusercontent.com/14938032/68523092-effc9000-02ee-11ea-929f-adfa2bfac71d.jpg)

Then I change the number of negative samples for relations to 2 and it works properly. Does it mean that, to implement the results (without aligned relations) as reported in the paper, I should just set the number of negative samples for relations to 0?

Thanks for your time!
Weixin.

**Comment by acharkq** [comment]
The reason for the reported error is the limited relation number, which is 3. The function failed to get 25 negative samples. Change the negative sample number for relation to 2 should be fine.

To implement the results without aligned relations, you should gently remove the `rel_align_loss` in line 135:
`loss = sum([align_loss, transe_loss, rel_align_loss, rule_loss])`
and line 138:
`loss = sum([align_loss, rel_align_loss, transe_loss])`
of `config.py`.

**Comment by DexterZeng** [comment]
Thank you for your reply!
